[docs] Deployment scenarios (#1364)
Azure/iotedge-lorawan-starterkit: docs/user-guide/deployment-scenarios.md

# Deployment Scenarios

> The StarterKit support a wide variety of deployments, and this article will

Reviewer: We use different wording for the starter kit in different places, maybe we should homogenize this?

highlight some of them.

- [Keep data on-premise](#keep-data-on-premise)
- [Redundancy with multiple concentrators and gateways](#redundancy-with-multiple-concentrators-and-gateways)
- [Deployment on Windows Server](#deployment-on-windows-server)

> - [Deployment of IoTEdge in the cloud](#deployment-of-iotedge-in-the-cloud)

Reviewer: ```suggestion
- [Deployment of IoT Edge in the cloud](#deployment-of-iotedge-in-the-cloud)
```

Other scenarios are supported, and combinations of the below scenario are possible.

## Keep data on-premise

It is possible to deploy Azure IoT Edge on your own hardware, and keep all data
in your local network or infrastructure. For example, by using a custom local
forwarding module, it is possible to route the data to a local queue or message
bus. In this case, the sensor data will never leave the on-premise network, and

> connection to Azure will only be used for managing the devices, or handling the

Reviewer: It's not clear to me what "Management flow" means. I think to be more accurate we would need to differentiate between how the user defines devices (which is how I would understand "management"), and which dependency the starter kit has "internally". I think the Facade function has more functionality than simply managing devices and dedup, it's there to fetch credentials, we use it for firmware upgrades etc.

Author: By "management flow" I mean exactly that: updating devices, handling dedup, etc... It is used to differentiate from actual "signal or telemetry data".

Author: @bastbu I updated the wording a little bit, does that make sense?

Reviewer: Hm, I'm still not sure if it makes sense that the Facade is only there to manage devices and for deduplication. But then again, it's such a detail, not worth changing. Will resolve the comment.

[deduplication](../adr/007_message_deduplication.md).

![Keep data on-premise](../images/scenarios/scenario-local-data.png)

## Redundancy with multiple concentrators and gateways

> In a setup where all sensors can reach at least one concentrator, it is possible

Reviewer: ```suggestion
In a setup where sensors can reach at least one concentrator, it is possible
```

> to get redundancy buy using at lest two concentrators per gateway, and at least

Reviewer: ```suggestion
to get redundancy by using at least two concentrators per gateway, and at least
```

> two gateways. In this case, the deduplication of the messages will happen in two

Reviewer: [nit] I think it would be more accurate to say that deduplication happens on two levels. The Function will only return whether a message is a duplicate, but it will not actually deduplicate the messages. This is still done on the gateway.

> two gateways. In this case, the deduplication of the messages will happen in two

Reviewer: Would it be worth mentioning that the user can configure how deduplication is handled?

places: on the gateway itself for concentrator deduplication, and in the Function
for gateway deduplication ([More info on deduplication strategies](../adr/007_message_deduplication.md))

> ![Redundancy](../images/scenarios/scenario-redundancy.png)

Reviewer: This is the only diagram where we have multiple concentrators but don't have the tag "LoRa Basics Station Concentrators". I found that description on the image useful, would it be worth adding it?

Author: added!

## Deployment on Windows Server

Some organizations have a limitation of only supporting Windows operating systems
on their infrastructure. In this case, it is still possible to use IoT Edge and
the StarterKit by using [*EFLOW*](https://docs.microsoft.com/en-us/azure/iot-edge/iot-edge-for-linux-on-windows?view=iotedge-2018-06),

> short for Edge For Linux On Windows. This allows you to run containerized Linux

Reviewer: The product team uses small letters for "for" and "on" on the product page.

```suggestion
short for "Edge for Linux on Windows". This allows you to run containerized Linux
```

workloads on Windows machines.

![Redundancy](../images/scenarios/scenario-eflow.png)

> ## Deployment of IoTEdge in the cloud

Reviewer: ```suggestion
## Deployment of IoT Edge in the cloud
```

It is also possible to run the IoT Edge modules directly in the cloud. Either a
Linux VM to run IoT Edge directly, or on a Windows VM with EFLOW.
In fact, the full end-to-end Continuous Integration pipeline for the StarterKit
runs on EFLOW on a VM in Azure.

> ![Redundancy](../images/scenarios/scenario-edge-in-cloud.png)

Reviewer: Should we mention that we can also mix-and-match these deployment scenarios or is that too obvious? It's possible to have multiple gateways in the cloud, one of them running on EFLOW and the other on Debian, etc etc.

Author: there is already a very short explanation, on line 11. Is that sufficient?

in your local network or infrastructure. For example, by using a custom local
forwarding module, it is possible to route the data to a local queue or message
bus. In this case, the sensor data will never leave the on-premise network, and
the connection to Azure will be used for managing the devices, handling the
[deduplication](../adr/007_message_deduplication.md) or managing [concentrator
updates](station-firmware-upgrade.md) for example.

![Keep data on-premise](../images/scenarios/scenario-local-data.png)

## Redundancy with multiple concentrators and gateways

In a setup where sensors can reach at least one concentrator, it is possible
to get redundancy by using at least two concentrators per gateway, and at least
two gateways. In this case, the deduplication of the messages will happen in two
levels: the Function will signal if a message is a duplicate, and the gateway
itself will handle the deduplication. It is possible to configure how
deduplication is handled and which strategy is used.  
([More info on deduplication strategies](../adr/007_message_deduplication.md))

![Redundancy](../images/scenarios/scenario-redundancy.png)

## Deployment on Windows Server

> Some organizations have a limitation of only supporting Windows operating systems

Reviewer 2: ```suggestion
Some organizations only support Windows Server operating systems
```

I would not see this as a 'limitation' and would include the 'Server' in it

on their infrastructure. In this case, it is still possible to use IoT Edge and
the Starter Kit by using [*EFLOW*](https://docs.microsoft.com/en-us/azure/iot-edge/iot-edge-for-linux-on-windows?view=iotedge-2018-06),

> short for "Edge for Linux on Windows". This allows you to run containerized Linux

Reviewer 2: ```suggestion
short for "Azure IoT Edge for Linux on Windows". This allows you to run containerized Linux
```

Official name is 'Azure IoT Edge for Linux on Windows'

workloads on Windows machines.

![Redundancy](../images/scenarios/scenario-eflow.png)

## Deployment of IoT Edge in the cloud

It is also possible to run the IoT Edge modules directly in the cloud. Either a
Linux VM to run IoT Edge directly, or on a Windows VM with EFLOW.
In fact, the full end-to-end Continuous Integration pipeline for the Starter Kit

> runs on EFLOW on a VM in Azure.

Reviewer 2: ```suggestion
also tests this scenario with an installation of EFLOW in a Windows Server VM in Azure.
```
